File: two_pointer/compare-strings-containing-backspaces.js
const backspace_compare = (str1, str2) => {
    let l = str1.length - 1;
    let r = str2.length - 1;

    while (l >= 0 && r >= 0) {
        let bl = 0;
        while (str1[l] === '#') {
            bl++
            l--;
        }

        l -= bl;
        bl = 0;

        let br = 0;
        while (str2[r] === '#') {
            br++
            r--;
        }

        r -= br;
        br = 0;

        if (str1[l] !== str2[r]) return false;
        r--;
        l--;
    }

    return true;
}

console.log(backspace_compare("xy#z", "xzz#")) // true
console.log(backspace_compare("xy#z", "xyz#")) // false
console.log(backspace_compare("xp#", "xyz##")) // true
console.log(backspace_compare("xywrrmp", "xywrrmu#p")) // true

File: two_pointer/python/compare-strings-containing-backspaces.py
def backspace_compare(str1, str2):
    str1_pt, str2_pt = len(str1) - 1, len(str2) - 1

    while str1_pt > -1 and str2_pt > -1:
        str1_backspaces = 0
        while str1[str1_pt] == "#":
            str1_backspaces += 1
            str1_pt -= 1

        while str1_backspaces > 0:
            str1_backspaces -= 1
            str1_pt -= 1

        str2_backspaces = 0
        while str2[str2_pt] == "#":
            str2_backspaces += 1
            str2_pt -= 1

        while str2_backspaces > 0:
            str2_backspaces -= 1
            str2_pt -= 1

        if str1[str1_pt] == str2[str2_pt]:
            str1_pt -= 1
            str2_pt -= 1
        else:
            return False

    return True if str1_pt == str2_pt else False


if __name__ == '__main__':
    print(backspace_compare("xy#z", "xzz#"))  # true
    print(backspace_compare("xy#z", "xyz#"))  # false
    print(backspace_compare("xp#", "xyz##"))  # true
    print(backspace_compare("xywrrmp", "xywrrmu#p"))  # true
